common/types/client.go

```diff
@@ -0,0 +1,31 @@
+package types
+
+import (
+	"context"
+	"math/big"
+)
+
+type Client[CHAINID any, ADDR comparable, BLOCK any, TX any, HASH comparable, TXRECEIPT any, EVENT any, EVENTOPS any] interface {
+	ChainID() (CHAINID, error)
```

Comment:
Why does this signature have an error? It looks like all the implementations return nil, and all of the callers ignore it.

Reply:
So the idea here is the following: as the part of `ChainClient`, `ChainId()` is expected to be the RPC call similar to other methods in the interface. This is how we implement it for other chains. The reason it doesn't return an error on the evm client currently is that it never calls the chain but rather returns the value initialised from node config. Imo it's not the behaviour one expects from the `ChainClient`: the consumer expects the real chain id call rather than getting one from the config (how do you even know the chain id is correct?). So in order to generalise this across all our implementations I add the RPC error to the return value.  
The next potential steps on the evm side would be to actually call the chain for chain id on the node startup, cache the value (validate against config possibly) and then return to consumers (same call of maybe `ChainIdCached()` that wouldn't require RPC error returned). Alternatively, consumers shouldn't use `ChainClient` for getting chain id at all and read it straight from config.

Reply:
Asking for the configured chain ID vs the client RPC chain ID are fundamentally different things that cannot be conflated.
If we need to expose the RPC variant with an error, then the returned errors _must_ be handled - and we should not be using that version in places that are currently using the configured chain ID.

Reply:
We already do Chain ID valiation on startup, so that you don't accidentally use a node in the pool that is for another chain ID. Please don't re-implement redundant checks in other places.

Reply:
I didn't plan for those updates in the scope of this PR. It's good we have it already 🙃 
My goal here is to draw the common denominator for the ChainClient across all the existing and future chains we plan to support. How about I keep `ChainId() (*big.Int, error)` as the generic RPC method and rename the existing `ChainId() *big.Int` into `ChainIdCached() *big.Int` (better name suggestions appreciated)?

Reply:
I'm not sure that we ever want a _cached_ value - what is the purpose of that?
We currently have an explicitly configured ID, and we have the ability to fetch via RPC. I think we should be preserving the existing behavior as-is during this conversion and any changes in this regard should be done separately.

Reply:
@sdrug: I think calling the ChainId() RPC call should only be done once at node startup, and then as @jmank88 said, validated with config to ensure we are connected to a correct node. But after that, every other call to client.ChainId() should be using the configured value.
A basic reason is, making RPC call for every time some1 needs a chain-id, will overload our node, and also DDOS the RPC nodes. Also introduce huge latency, all of which are unacceptable.

However I think in future, there can be an explicit param, or a different method, which a client exposes, to force an RPC call to get ChainId(). But lets add that only if a usecase arises that needs to force an RPC call.

Reply:
I also notice another thing. Lot of code gets chainId from the client, and not the config.
Ideally, all callers should get chainId from config, and should call client.ChainId() only if they actually want to make an RPC call. The Client interface exposed methods should only be for RPC calls.

Anyways, trying to address this behavior sort of is out of scope of this PR.
@sdrug : Could you just change ChainId() call to always only return internal configured value for now? If we want to change that later, we would need some refactoring, an change all calling code too, which isn't something for this PR.

Reply:
I agree with you on both points. I didn't mean to overload the node with RPC calls every time. My take was that returning configured chain id rather then calling the RPC is not the behaviour one would expect from ChainClient. And yeah, in this case I'd expect consumers to rather call config directly but not through ChainClient.

Reply:
Yeah I think clear naming is the first step, and then we have to untangle the existing usages. I think that something simple like this would be sufficient:
```suggestion
	ConfiguredChainID() CHAINID
	ChainID() (CHAINID, error)
```
But maybe something like this could make it a little simpler, and forward-compatible with any other fields we want to add in the future:
```suggestion
	Config() Config[CHAINID]
	ChainID() (CHAINID, error)
```
```go
type Config[CHAINID any] struct {
	ChainID CHAINID
	// add more fields later w/o breaking the API
}
```

Reply:
I like the first version 👍 
For other config values I'd rather encourage to use config interfaces and keep the ChainClient as clean as possible.